Fix to refuse an array from a different device in `Variable.array`

okapies opened the pull request:
This patch adds `Device.is_compatible_array()` method and fixes to reject an array coming from a non-compatible device in the setter of `Variable.array`.

Fixes #6058. Related to #6116. Blocked by ~#7487~.

okapies commented:
Jenkins, test this please.

okapies commented on chainer/variable.py:
```diff
@@ -1792,18 +1808,19 @@ def initialize(self, shape):
+        # (generate_array produces numpy array even if device is Intel64)
+        # TODO(niboshi): This could be done in generate_array().
+        if isinstance(device, intel64.Intel64Device):
+            data = device.send_array(data)
```
It converts `data` to `ideep.mdarray` before `self.array = data` because now `array()` rejects a numpy array if `device is Intel64Device`.

okapies commented:
Jenkins, test this please.

niboshi commented on chainer/variable.py:
```diff
@@ -1143,21 +1160,22 @@ def _to_device(self, device, allow_unchaining):
-            return
-
-        if backend.get_device_from_array(arr) == device:
+        if arr is None or old_device == device:
```
How about checking `old_device == device` earlier?

okapies replied:
It depends on the conclusion of #6116 because `Parameter` expects `self._device` has a valid configuration even if `arr is None`. We can move `old_device == device` independently but I’d like to wait until the discussion is finished.

okapies replied:
I'd like to leave it because #6116 has not yet been concluded.

niboshi replied:
OK.

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -285,6 +286,35 @@ def test_uninitialized(self):
         assert a.xp is np
         assert a._has_chainerx_array is False
 
+    def check_array(self, xp, x, y, requires_grad=True):
```
Some suggestions for the test:
* We should test for every possible pairs of devices. How about making a separate `TestCase` and parameterizing using two `inject_backend_config`s?
* The code inside `pyteset.raises` should be minimal. Enclosing the entire call of `check_array` could make the test pass unexpectedly by an assertion failure from some unexpected place.

niboshi commented on chainer/variable.py:
```diff
@@ -511,6 +511,10 @@ class Variable(object):
     # TODO(hvy): Avoid modifying this variable from outside this class.
     _chainerx_fallback_array = None
 
+    _data = None  # type: tp.List[tp.Optional[types.NdArray]]
```
I think it's better to place these attribute declarations before ChainerX-specific attributes.

niboshi commented on chainer/variable.py:
```diff
@@ -511,6 +511,10 @@ class Variable(object):
     # TODO(hvy): Avoid modifying this variable from outside this class.
     _chainerx_fallback_array = None
 
+    _data = None  # type: tp.List[tp.Optional[types.NdArray]]
```
Please remove type hints from internal interfaces and code.

okapies replied:
Hmm, it is required to remove type errors in the user-facing methods. How do you think about this kind of cases?

niboshi replied:
Could you elaborate that? I thought internal attributes should not affect users' typing checks.

okapies replied:
If we specify types for arguments, mypy try to check if all variables (including the arguments) in the function scope are sound. It doesn't affect the user's type checking, but leaving these errors are not convenient because the type error output is useful for adding new type hints even if CI doesn't run type checking.

niboshi replied:
Mmm, I don't still understand. What do you mean by it doesn't affect user's type checking? It sounds like it won't cause any errors.
Could you post actual commands?

okapies replied:
I tend to run `mypy chainer` in the `chainer` directory when adding type hints to Chainer for inspecting *type bugs*.

niboshi replied:
Is that for checking chainer's internal type consistency?
As discussed previously (#6146), we're not going to guarantee such check would succeed.

okapies replied:
Yes, but sometimes, in my experience, the types of interface method are not obvious from the documentation. For example, I noticed #4074 independently because of the mypy error. In other case, the specification of `Initializer` [allows a scalar value](https://docs.chainer.org/en/stable/reference/initializers.html) but it is not explicitly documented in [some APIs](https://github.com/chainer/chainer/blob/7e4f072c6d42cd345ac216cfdf353bce891140e0/chainer/links/connection/linear.py#L34-L38). I noticed `InitializerSpec` should contain `ScalarValue` when running mypy.

I agree that non-public interfaces and variables should prevent from using type hints, but minimum usage is very useful for detecting problems described like above. How do you think?

niboshi replied:
So do you mean you want to restart using mypy?
It's an option of course, but the team should agree with that first, as the current concensus is the opposite.

By the way, `L.Linear` docs has a link to initializers, so I'd say it's documented.

okapies replied:
Ok, I'm honestly wondering what means "use type hints only in the public interfaces," but I understood that the type hints basically should not be used in Chainer. I'll remove this hint and create a PR for the documentations. Thanks for the discussion.

niboshi commented:
Jenkins, test this please

niboshi commented:
It seems some tests failed.

okapies commented:
Thanks for the catch. I'll check it.

niboshi commented:
@okapies How is it going?

okapies commented:
Jenkins, test this please.

okapies commented:
I remember that what are problems. For example, the following failure is *reasonable* because `Intel64Device` is a different instance from `CpuDevice`. Should we accept such the case? @niboshi 

```
ValueError: The specified array is invalid. Variable is configured against @intel64 but it is on @numpy.
```

niboshi commented:
I see. It should be accepted because `Intel64Device` handles `numpy.ndarray` as well.
This seems to be a difficult problem.
You could use `Device.supported_array_types` to check the array type, but it's not enough because `ChainerxDevice` with `cuda:0` should not accept `chainerx.ndarray` on `cuda:1`.

niboshi commented:
Hmm, this fix looks a bit too adhoc.
`ChainerxDevice` and other devices should behave similarly regarding this point.
One possible solution would be to introduce somethig like `Device.is_arrays_compatible(arrs)` (and perhaps remove `Device.supported_array_types`).

okapies commented:
Added `Device.is_compatible_array` method to check if the specified array is compatible with the device. I also found that some tests try to do `variable.array = None` as a result of invoking `copy`. I changed to accept these cases but it should be solved in (3.) of #6116.

okapies commented:
I found that `tests/chainer_tests/test_variable.py` fails in single GPU environment (maybe this PR add the check for `Device`). Is there good workaround for it?

okapies commented:
Jenkins, test this please.

okapies commented:
I got the same errors in `master` branch. I'll post an issue about it. https://github.com/chainer/chainer/pull/6072#issuecomment-496884300

okapies commented:
Jenkins, test this please.

okapies commented:
@niboshi All issues are solved I think. Could you please check it?

niboshi commented on chainer/variable.py:
```diff
@@ -1185,15 +1203,15 @@ def _to_device(self, device, allow_unchaining):
             self._chainerx_name = self._node.name
 
+        # Note: It affects the whole behavior of Variable's methods.
+        #       We must place them in this place.
```
Hmm, I don't get what this comment is trying to convey.

okapies replied:
`self._device` and `self._has_chainerx_array` is referenced in many places both explicitly and implicitly. Reassining them leads to change the behavior of some methods and properties. For example, `grad_var = self.grad_var` doesn't work as expected if we place them before it.

niboshi replied:
I don't see any particular reason to make such comment to these lines.
It's in fact very common that timing of attribute assignment matters.

okapies replied:
Hmm, OK. I'll remove it.

okapies commented on chainer/variable.py:
```diff
@@ -849,27 +851,36 @@ def array(self):
+                new_device = backend.get_device_from_array(d)
+                raise ValueError(
+                    "The specified array is incompatible. Variable is "
+                    "configured against {} but it is on {}.".format(
```
Fixed. But can we use multiple lines in the exception message?

niboshi replied:
Yes. Multiline message is used in many places.

okapies replied:
Ok, I'll fix.

okapies commented on chainer/variable.py:
```diff
@@ -951,6 +962,7 @@ def _set_grad_without_check(self, g):
 
     @property
     def grad(self):
+        # type: () -> tp.Optional[types.NdArray]
```
Moved to #7400.

niboshi commented on chainer/_backend.py:
```diff
@@ -105,3 +105,16 @@ def use(self):
          """
         pass
+
+    def is_array_compatible(self, array):
```
`is_array_compatible` needs its own tests.
Could you separate a PR?

okapies replied:
I see. Moved to #7487.

niboshi commented on chainer/variable.py:
```diff
@@ -858,18 +860,27 @@ def array(self):
+            old_device = self.device
+            if not old_device.is_array_supported(d):
+                new_device = backend.get_device_from_array(d)
+                raise ValueError(
```
This check won't be triggered if `self._data[0] is None`. I think it should.

okapies replied:
`d_old is None` can be `True` as a *normal* case when unpickle a `Parameter`.

1. A `Parameter` instance is created in uninitialized state
2. The unpickler calls `__reduce__` and `_recover_parameter`
3. `_recover_parameter` assigns recovered data to `array` property

niboshi replied:
Yes, that's true. Is it valid to assign an array of different device in above cases?

okapies replied:
Hmm, do you mean the case when `self._data[0] is None` and `self.device is not None`? In that case, we should check `old_device` instead of `d_old`.

okapies replied:
Ah, I just remember that `self.device` (= `old_device`) returns `CpuDevice` if `self._data[0] is None`.

niboshi replied:
Can you add a TODO? We should check the device in such cases anyway.

okapies replied:
I tried to fix the case like the following:

1. Disable caching in the getter of `device` property when `self._data[0] is None`
    - It means `self._device` is always `None` before `self._data[0]` is assigned once
2. Check `_device` field directly in the setter of `array` property instead of accessing `device` property

I'm running the tests...

niboshi replied:
Hmm, disabling cache might have performance implication.

okapies replied:
I think it is not a major issue because:

1. `device` will not be called frequently when `Variable`/`Parameter` is uninitialized
2. Creating `CpuDevice` instance incurs relatively low overhead

niboshi replied:
Is that really true?

>1. `device` will not be called frequently when `Variable`/`Parameter` is uninitialized

I don't have a total picture of how `device` attribute is used by Chainer itself or users.
The only thing I can say is `device` is meant to be lightweight.

>2. Creating `CpuDevice` instance incurs relatively low overhead

We have made `device` attribute cached because there was a performance issue.
I don't think the overhead for creating `CpuDevice` is negligible.

okapies replied:
Hmm. In this sense, maybe we need another flag like `self._initialized` (it is the same as `self._is_cached_real_device`) or something to distinct `self._device == CpuDevice()` is a real value. How do you think?

okapies replied:
Another idea is to change the spec of `device` as returning `None` when `Variable` is uninitialized.

niboshi replied:
Yes, we need some change as you described above, and that should be in another PR.
That's why I proposed a TODO.

okapies replied:
I see. Fixed and added some tests for unpickling uninitialized `Parameter`.

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -1420,6 +1420,71 @@ def test_bool_inconvertible(self):
+
+        should_fail = (
+            (
+                not device_x == device_y
```
`device_x != device_y`

okapies replied:
Fixed (it was to handle #6335).

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -1420,6 +1420,71 @@ def test_bool_inconvertible(self):
+        should_fail = (
+            (
+                not device_x == device_y
+                and not device_x == intel64.Intel64Device) or
```
Do you mean `isinstance`?
Currently this expression should always evaluate to `True`.

okapies replied:
Good catch, fixed.

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -1420,6 +1420,71 @@ def test_bool_inconvertible(self):
+            # should not accept an array from a different device
+            # or `x` or `y` is a chainerx array which requires gradient
+            with pytest.raises(ValueError):
+                v.data = y
```
`v.array = y`

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -1420,6 +1420,71 @@ def test_bool_inconvertible(self):
+                v.data = y
+        else:
+            # Should succeed
+            v.data = y
```
Ditto.

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -1420,6 +1420,71 @@ def test_bool_inconvertible(self):
+        else:
+            # Should succeed
+            v.data = y
+            xp_x.testing.assert_array_equal(v.data, y)
```
Ditto.

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -1420,6 +1420,71 @@ def test_bool_inconvertible(self):
+
+    def _gen_array(self, device, shape, requires_grad):
+        if shape is None:
+            a = None
```
return `None`

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -1420,6 +1420,71 @@ def test_bool_inconvertible(self):
+    }))
+@testing.backend.inject_backend_tests(None, _set_data_backend_params)
+@testing.backend.inject_backend_tests(None, _set_data_backend_params)
+class TestVariableSetData(unittest.TestCase):
```
`TestVariableSetArray`

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -1420,6 +1420,84 @@ def test_bool_inconvertible(self):
+            if isinstance(x, intel64.mdarray):
+                should_fail = not isinstance(y, (np.ndarray, intel64.mdarray))
+            else:
+                should_fail = type(x) != type(y)
```
`(type(x), type(y)) == (numpy.ndarray, intel64.mdarray)`
Shouldn't this case succeed?

niboshi commented on tests/chainer_tests/test_variable.py:
```diff
@@ -1854,6 +1932,43 @@ def test_copy_to_initialize(self):
         x.initialize((3, 2))
         assert x.data is y.data
 
+    def test_set_array_numpy(self):
```
Please separate those tests and use `inject_backend_tests` just like `TestVariableSetArray`.

niboshi commented:
Could you separate unrelated changes?
Every time non-trivial change is made in this PR, I have to check every single line.

okapies commented:
@niboshi I'm very sorry to late response. Could you check it?
